crates/evm/src/lib.rs

```diff
@@ -31,6 +31,132 @@ pub mod system_calls;
 /// test helpers for mocking executor
 pub mod test_utils;
 
+/// Builder for creating an EVM with a database and environment.
+#[derive(Debug)]
+pub struct RethEvmBuilder<DB: Database, EXT> {
+    /// The database to use for the EVM.
+    db: DB,
+    /// The environment to use for the EVM.
+    env: Option<Box<Env>>,
+    /// The external context for the EVM.
+    external_context: EXT,
+}
+
+impl<DB, EXT> RethEvmBuilder<DB, EXT>
+where
+    DB: Database,
+{
+    /// Create a new EVM builder with the given database.
+    pub const fn new(db: DB, external_context: EXT) -> Self {
+        Self { db, env: None, external_context }
+    }
+
+    /// Set the environment for the EVM.
+    pub fn with_env(mut self, env: Box<Env>) -> Self {
+        self.env = Some(env);
+        self
+    }
+
+    /// Build the EVM with the given database and environment.
+    pub fn build<'a>(self) -> Evm<'a, EXT, DB> {
+        let mut builder =
+            EvmBuilder::default().with_db(self.db).with_external_context(self.external_context);
+        if let Some(env) = self.env {
+            builder = builder.with_env(env);
+        }
+
+        builder.build()
+    }
+
+    /// Build the EVM with the given database and environment, using the given inspector.
+    pub fn build_with_inspector<'a, I>(self, inspector: I) -> Evm<'a, I, DB>
+    where
+        I: GetInspector<DB>,
+        EXT: 'a,
+    {
+        let mut builder =
+            EvmBuilder::default().with_db(self.db).with_external_context(self.external_context);
+        if let Some(env) = self.env {
+            builder = builder.with_env(env);
+        }
+        builder
+            .with_external_context(inspector)
+            .append_handler_register(inspector_handle_register)
+            .build()
+    }
+}
+
+/// Trait for configuring an EVM builder.
+pub trait ConfigureEvmBuilder {
+    /// The type of EVM builder that this trait can configure.
+    type Builder<'a, DB: Database>: EvmFactory;
+}
+
+/// Trait for configuring the EVM for executing full blocks.
+#[auto_impl::auto_impl(&, Arc)]
+pub trait EvmFactory: ConfigureEvmEnv {
+    /// Associated type for the default external context that should be configured for the EVM.
+    type DefaultExternalContext<'a>;
+
+    /// Provides the default external context.
+    fn default_external_context<'a>(&self) -> Self::DefaultExternalContext<'a>;
+
+    /// Returns new EVM with the given database
+    ///
+    /// This does not automatically configure the EVM with [`ConfigureEvmEnv`] methods. It is up to
+    /// the caller to call an appropriate method to fill the transaction and block environment
+    /// before executing any transactions using the provided EVM.
+    fn evm<DB: Database>(&self, db: DB) -> Evm<'_, Self::DefaultExternalContext<'_>, DB> {
```

Comment:
I'd like this to consume self, because this is just the type that collapses to an evm

Reply:
fixed